example/wasapi.dart

```diff
@@ -83,8 +83,68 @@ void main() {
   // Initialize COM
   check(CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED));
 
-  // Retrieve the default audio output device.
+  // Retrieve the list of available audio output devices.
   final pDeviceEnumerator = MMDeviceEnumerator.createInstance();
+  final ppDevices = calloc<Pointer<COMObject>>();
+  check(pDeviceEnumerator.enumAudioEndpoints(
+    0, // dataflow: rendering device
+    1, // device state: only enumerate active device
+    ppDevices));
+
+  // Get the number of available audio output devices.
+  final pDevices = IMMDeviceCollection(ppDevices.cast());
+  final pcDevices = calloc<Uint32>();
+  check(pDevices.getCount(pcDevices.cast()));
+  final deviceCount = pcDevices.value;
+  print("$deviceCount audio device(s) detected:");
+
+  // Print available audio output devices
+  for (var i = 0; i < deviceCount; i++) {
+    // Get audio device from the device collection.
+    final ppEndpoint = calloc<Pointer<COMObject>>();
+    check(pDevices.item(i, ppEndpoint.cast()));
+    final pEndpoint = IMMDevice(ppEndpoint.cast());
+
+    // Retrieve the current device id
+    final idPtr = calloc<Pointer<Utf16>>();
+    pEndpoint.getId(idPtr.cast());
+    final id = idPtr.value.toDartString();
+    free(idPtr);
+
+    // Retrieve the current device properties.
+    final ppProps = calloc<Pointer<COMObject>>();
+    check(pEndpoint.openPropertyStore(
+        STGM.STGM_READ, // Storage-access mode: read
+        ppProps.cast()));
+
+    // Build property key to get device friendly name.
+    final pProps = IPropertyStore(ppProps.cast());
+    final guid = calloc<GUID>();
+    CLSIDFromString(PKEY_DeviceInterface_FriendlyName.toNativeUtf16(), guid);
+    final pKey = calloc<PROPERTYKEY>()
+      ..ref.fmtid = guid.ref
+      ..ref.pid = 2;
```

Comment:
The annoying thing about getting the property is that we need to use the correct `GUID` **AND** Property ID (`pid`). Specifying both every `pProps.getValue` looks very error-prone to me.  

Is there a way to define these `GUID` and `pid` together? If so, how should I implement it? An enum in some place perhaps?

Reply:
Just for reference, the C# binding for `win32` seems to store these guids inside a static class and couple the guid and pid together via a C# attribute called `Constant`.

I do think we should do something like this as well.
![image](https://github.com/dart-windows/win32/assets/9136829/015ff73b-5b18-4f83-b740-4375711b51e7)

Reply:
Is this screenshot from ILSpy? Just to clarify, it doesn't represent the C# bindings for the `PROPERTYKEY` properties. ILSpy provides a glimpse into the API metadata being introspected. It's up to language projection authors to decide how to utilize this information.

[Constant](https://github.com/microsoft/win32metadata/blob/bcce701782815f62c550465072c66592e330ff35/generation/WinSDK/manual/Metadata.cs#L66-L71) attribute is part of the [win32metadata](https://github.com/microsoft/win32metadata) and it includes values of the struct initializers as a comma-separated string. In this case, `PROPERTYKEY`'s `fmtid` and `pid` values.

For example, you can check out the Rust projection of these properties [here](https://github.com/microsoft/windows-rs/blob/02db74cf5c4796d970e6d972cdc7bc3967380079/crates/libs/sys/src/Windows/Win32/Devices/Properties/mod.rs), which is auto-generated from the win32metadata by utilizing the values inside the `Constant` attribute.

In Rust, destructors can be implemented for any struct through the Drop trait. This is typically done to ensure the automatic cleanup of resources when an instance of the struct goes out of scope. Such a mechanism enables the creation of constants, like the ones illustrated. However, Dart does not support this. Consequently, manual resource deallocation is required, rendering it impractical to create constants or enums for properties of this nature in Dart.

BTW, there is a better way to create these properties in your code.

Instead of doing this:
```dart
  final guid = calloc<GUID>();
  CLSIDFromString(PKEY_DeviceInterface_FriendlyName.toNativeUtf16(), guid);
  final pKey = calloc<PROPERTYKEY>()
    ..ref.fmtid = guid.ref
    ..ref.pid = 2;

  ...

  free(guid);
  free(pKey);
```

You can do this:
```dart
  final PKEY_DeviceInterface_FriendlyName = calloc<PROPERTYKEY>()
    ..ref.fmtid.setGUID('{026e516e-b814-414b-83cd-856d6fef4822}')
    ..ref.pid = 2;

  ...

  // Free the memory allocated for the struct when it is no longer needed
  free(PKEY_DeviceInterface_FriendlyName);
```

Reply:
Hi, I've updated the code as your suggestion. And yes, it is a similar program to ILSpy called dnSpy. I see, it would make sense for Rust, but I believe my point still stand. Manually creating `PROPERTYKEY` by setting GUID and pid is sounds very error-prone to me, especially when the program deals with many properties.  

I would love to see a way to at least couple these GUIDand its pid in one place. I was thinking about creating an enum that consists only of the GUID string and PID and then let the user construct the `PROPERTYKEY` by themselves with these values enum values.

Reply:
We can consider adding static methods into the `PROPERTYKEY` struct:

```dart
base class PROPERTYKEY extends Struct {
  external GUID fmtid;

  @Uint32()
  external int pid;

  static Pointer<PROPERTYKEY> DeviceInterface_FriendlyName({
    Allocator allocator = calloc,
  }) =>
      allocator<PROPERTYKEY>()
        ..ref.fmtid.setGUID('{026e516e-b814-414b-83cd-856d6fef4822}')
        ..ref.pid = 2;

  ...
}
```